src/wms/src/main/java/org/geoserver/wms/decoration/LegendDecoration.java

```diff
@@ -293,4 +313,55 @@ private String findTitle(Layer layer, Catalog catalog) {
         public BufferedImage legend;
         public GetLegendGraphicRequest request;
     }
+
+    private LegendInfo resolveLegendInfo(
+            LegendInfo legendInfo, GetLegendGraphicRequest request, StyleInfo context) {
+        if (legendInfo == null) {
+            return null; // not available
+        }
+        String onlineResource = legendInfo.getOnlineResource();
+        String baseUrl = request.getBaseUrl();
+        if (onlineResource == null) {
+            return null;
+        }
+        URL url = null;
+        try {
+            URI uri = new URI(onlineResource);
+            GeoServerResourceLoader resources = wms.getCatalog().getResourceLoader();
+            if (uri.isAbsolute()) {
+                if (baseUrl != null && onlineResource.startsWith(baseUrl + "styles/")) {
+                    // convert relative to styles durectory
+                    onlineResource = onlineResource.substring(baseUrl.length() + 7);
+                } else {
+                    legendInfo.setOnlineResource(new URL(onlineResource).toString());
+                    return legendInfo; // an actual external graphic reference
+                }
+            } else {
+                // not absolute, try relative to the style if available, otherwise search
+                // in the styles directory
+                if (context != null) {
+                    GeoServerDataDirectory dd = new GeoServerDataDirectory(resources);
+                    Resource styleParentResource = dd.get(context);
+                    if (styleParentResource != null
+                            && styleParentResource.getType() == Resource.Type.DIRECTORY) {
+                        url =
+                                URLs.fileToUrl(
+                                        new File(
+                                                styleParentResource.dir().getCanonicalPath(),
+                                                onlineResource));
+                    }
+                }
+            }
+            if (url == null) {
+                File styles = resources.findOrCreateDirectory("styles");
+                URL base = URLs.fileToUrl(styles);
+                url = new URL(base, onlineResource);
+            }
+        } catch (Exception e) {
+            return null; // Do not try this online resource
+        }
+        legendInfo.setOnlineResource(url.toExternalForm());
```

Comment:
Hmm... LegendInfo is part of the catalog configuration, are we sure that it can be modified just like this without side effects? Is it wrapped in a ModificationProxy at runtime? If so, unless it gets saved, there are no problems, otherwise, best to clone it before changing it.

Reply:
fixed